packages/fetchai/connections/http/connection.py

```diff
@@ -0,0 +1,416 @@
+            to=self.address,
+            sender=self._client_id,
+            protocol_id=self.protocol_id,
+            context=context,
+            message=msg_bytes,
+        )
+        return envelope
+
+    def send(self, envelope: Envelope) -> None:
+        """
+        Send the envelope in_queue.
+
+        :param envelope: the envelope
+        :return: None
+        """
+        assert self.in_queue is not None, "Input queue not initialized."
+        assert self.loop is not None, "Loop not initialized."
+        asyncio.run_coroutine_threadsafe(self.in_queue.put(envelope), self.loop)
+
+    def _send_response(self):
+        """Pass the response from out-bounded Envelope back to the front-end client.
+
+        Currently, the response will be written back in the cmd terminal.
+        """
+        # self.wfile.write(self.path.encode())
+        pass
+
+    def connect(self):
+        """
+        Connect.
+
+        Upon HTTP Channel connection, kickstart the HTTP Server.
+        """
+        with self.lock:
+            if self.stopped:
+                self.stopped = False
+                # self.thread.start()
+                # Initializing the internal HTTP server
+                try:
+                    self.httpd = HTTPServer(
+                        (self.host, self.port), HTTPHandlerFactory(self)
+                    )
+                    # self.httpd = HTTPServer((self.host, self.port), handler)
+                    logger.info(f"HTTP Server has connected to port {self.port}.")
+                    self.httpd.serve_forever()
+                except OSError:
+                    logger.error(
+                        f"{self.host}:{self.port} is already in use, please try another Socket."
+                    )
+
+    def disconnect(self) -> None:
+        """
+        Disconnect.
+
+        Shut-off the HTTP Server and then HTTP Channel.
+        """
+        with self.lock:
+            if not self.stopped:
+                self.httpd.shutdown()
+                logger.info(f"HTTP Server has shutdown on port {self.port}.")
+                self.stopped = True
+                # self.thread.join()
+
+
+def HTTPHandlerFactory(channel: HTTPChannel):
+    class HTTPHandler(BaseHTTPRequestHandler):
+        def __init__(self, *args, **kwargs):
+            self._channel = channel
+            super(HTTPHandler, self).__init__(*args, **kwargs)
+
+        def do_HEAD(self):
+            self.send_response(200)
+            self.send_header("Content-type", "text/html")
+            self.end_headers()
+
+        def do_GET(self):
+            """Respond to a GET request."""
+            self.send_response(200)
+            self.send_header("Content-type", "text/html")
+            self.end_headers()
+
+            parsed_path = urlparse(self.path)
+            url = parsed_path.path
+            # url = parsed_path.geturl()
+            param = parse_qs(parsed_path.query)
+            param = json.dumps(param)
+
```

Comment:
at this point I'd have something like:
```
is_valid_request, response_code = verifier.verify("GET", url, param)
if is_valid_request:
    # create envelope
    # send envelope into agent
    # wait for response envelope
   if # response envelope arrives within timeout window:
      # serve its payload
   else:
    response_code = 408 # request timeout
    self.send_response(response_code)
    self.send_header("Content-type", "text/html")
    self.end_headers()
else:
    self.send_response(response_code)
    self.send_header("Content-type", "text/html")
    self.end_headers()
```

Reply:
Fixed, I'm still using the function `_send()`, which I renamed to `_process()` in HTTPHandler.